src/storage/src/storage_state.rs

```diff
@@ -831,6 +836,50 @@ impl<'w, A: Allocate> Worker<'w, A> {
         }
     }
 
+    /// Pumps latest status updates from the buffer shared with operators and
+    /// reports any updates that need reporting.
+    pub fn report_status_updates(&mut self, response_tx: &ResponseSender) {
+        let mut to_report = Vec::new();
+
+        // Pump updates into our state and stage them for reporting.
+        if self.storage_state.shared_status_updates.borrow().len() > 0 {
+            for shared_update in self.storage_state.shared_status_updates.take() {
+                let id = shared_update.id;
+
+                to_report.push(shared_update.clone());
+
+                self.storage_state
+                    .latest_status_updates
+                    .insert(id, shared_update.clone());
+                self.storage_state
+                    .reported_status_updates
+                    .insert(id, shared_update);
+            }
+        }
+
+        // We reset what we have reported when a new client/controller connects,
+        // such that we can re-send our latest status updates.
+        //
+        // And here we report any status where our latest known status differs
+        // from what has been reported.
+        for (id, latest_update) in self.storage_state.latest_status_updates.iter() {
+            let reported_update = self.storage_state.reported_status_updates.get(id);
+            if let Some(reported_update) = reported_update {
+                if reported_update == latest_update {
+                    continue;
+                }
+            }
+            to_report.push(latest_update.clone());
+            self.storage_state
+                .reported_status_updates
+                .insert(id.clone(), latest_update.clone());
+        }
+
+        if to_report.len() > 0 {
+            self.send_storage_response(response_tx, StorageResponse::StatusUpdates(to_report));
```

Comment:
A place that's going to conflict with https://github.com/MaterializeInc/materialize/pull/31261 (fyi @petrosagg)